File: VB/dxCustomButtonEdit/CustomButtonEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports System.Drawing
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.LookAndFeel
Imports System.Windows.Forms

Namespace dxCustomButtonEdit
	<ToolboxItem(True)>
	Public Class CustomButtonEdit
		Inherits ButtonEdit

		 Shared Sub New()
			RepositoryItemCustomButtonEdit.RegisterCustomButtonEdit()
		 End Sub
		 Public Sub New()
		 End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemCustomButtonEdit.EditorName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public Shadows ReadOnly Property Properties() As RepositoryItemCustomButtonEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemCustomButtonEdit)
			End Get
		End Property
	End Class

	Public Class TextEditorButton
		Inherits EditorButton

	End Class

	<UserRepositoryItem("RegisterCustomButtonEdit")>
	Public Class RepositoryItemCustomButtonEdit
		Inherits RepositoryItemButtonEdit

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content), RefreshProperties(RefreshProperties.All)>
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return EditorName
			End Get
		End Property
		Shared Sub New()
			RegisterCustomButtonEdit()
		End Sub
		Public Sub New()

		End Sub
		Friend Const EditorName As String = "CustomButtonEdit"

		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemCustomButtonEdit = TryCast(item, RepositoryItemCustomButtonEdit)
				If source Is Nothing Then
					Return
				End If
			Finally
				EndUpdate()
			End Try
		End Sub

		Public Shared Sub RegisterCustomButtonEdit()
            Dim _image As Image = Nothing

            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(EditorName, GetType(CustomButtonEdit), GetType(RepositoryItemCustomButtonEdit), GetType(ButtonEditViewInfo), New CustomButtonEditPainter(), True, _image, GetType(DevExpress.Accessibility.ButtonEditAccessible)))
        End Sub
	End Class

	Public Class CustomButtonEditPainter
		Inherits ButtonEditPainter

		Protected Overrides Sub DrawButton(ByVal viewInfo As ButtonEditViewInfo, ByVal info As EditorButtonObjectInfoArgs)
			If info.Button.Tag IsNot Nothing AndAlso info.Button.Tag.ToString() = "Custom" Then
                Dim look = New UserLookAndFeel(viewInfo.OwnerEdit)
                Dim painter As New SkinEditorButtonPainter(look)
				look.SkinName = look.ActiveSkinName
				look.UseDefaultLookAndFeel = False
				look.SkinMaskColor = info.Button.Appearance.BackColor
				painter.DrawObject(info)
			Else
				MyBase.DrawButton(viewInfo, info)
			End If
		End Sub
	End Class
End Namespace

File: CS/dxCustomButtonEdit/CustomButtonEdit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using System.Drawing;
using DevExpress.XtraEditors.Controls;
using DevExpress.LookAndFeel;
using System.Windows.Forms;

namespace dxCustomButtonEdit
{
    [ToolboxItem(true)]
    public class CustomButtonEdit : ButtonEdit
    {
         static CustomButtonEdit() 
        {
            RepositoryItemCustomButtonEdit.RegisterCustomButtonEdit();
        }
         public CustomButtonEdit() 
         {
         }
        public override string EditorTypeName 
        {
            get { return RepositoryItemCustomButtonEdit.EditorName; } 
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemCustomButtonEdit Properties
        {
            get { return base.Properties as RepositoryItemCustomButtonEdit; } 
        }
    }

    public class TextEditorButton : EditorButton{
        
    }

    [UserRepositoryItem("RegisterCustomButtonEdit")]
    public class RepositoryItemCustomButtonEdit : RepositoryItemButtonEdit
    {
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content), RefreshProperties(RefreshProperties.All)]
        public override string EditorTypeName
        {
            get { return EditorName; }
        }
        static RepositoryItemCustomButtonEdit() 
        { 
            RegisterCustomButtonEdit(); 
        }
        public RepositoryItemCustomButtonEdit() {

        }
        internal const string EditorName = "CustomButtonEdit";

        public override void Assign(RepositoryItem item)
        {
            BeginUpdate();
            try
            {
                base.Assign(item);
                RepositoryItemCustomButtonEdit source = item as RepositoryItemCustomButtonEdit;
                if (source == null) return;
            }
            finally
            {
                EndUpdate();
            }
        }
         
        public static void RegisterCustomButtonEdit()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(EditorName, typeof(CustomButtonEdit),
                typeof(RepositoryItemCustomButtonEdit), typeof(ButtonEditViewInfo), 
                new CustomButtonEditPainter(), true, null, typeof(DevExpress.Accessibility.ButtonEditAccessible)));
        }
    }

    public class CustomButtonEditPainter : ButtonEditPainter
    {
        protected override void DrawButton(ButtonEditViewInfo viewInfo, EditorButtonObjectInfoArgs info)
        {
            if (info.Button.Tag != null && info.Button.Tag.ToString() == "Custom")
            {
                var look = new UserLookAndFeel(viewInfo.OwnerEdit);                
                look.SkinName = look.ActiveSkinName;
                look.UseDefaultLookAndFeel = false;
                look.SkinMaskColor = info.Button.Appearance.BackColor;
                SkinEditorButtonPainter painter = new SkinEditorButtonPainter(look);
                painter.DrawObject(info);
            }
            else
                base.DrawButton(viewInfo, info);
        }
    }
}
